Fix #787 and add Unicode to ASCII function

samitainio opened the pull request:
Fix #787
- Uses regex to pick up the hostnames/domains from the "Received: from" headers.

Unicode to ASCII function
- Spam messages more often than not contain junk text as unicode characters in the headers. The "from" and "subject" headers being the most common ones. Before this change the script would error on such emails or sometimes replace the unicode characters with questionmarks "?".
- Function takes argument as an input and then encodes it in ascii while ignoring any malformed data. It then returns an ASCII string without the unicode characters.
- Currently implemented for "from" and "subject" handling.

Rafiot commented:
Can you give a sample of unicode not supported by MISP? Because we have email subjects with emojis, for sure.

The from field should not be a problem either, afaik.

samitainio commented:
@Rafiot 

I came across this one 💋 (U+1F48B) in one of my spams which errors when I try to add it with the original code. That one was parsed from the "from" headers and seems to be erroring only there for some reason.

I tested it now with different variations and all of these give error for me:

```
From: 💋 some text 💋 <user@domain.com>
From:  some text 💋 <user@domain.com>
From: 💋 some text <user@domain.com>
From:  some 💋 text <user@domain.com>
```

This one went through without errros so I uhm just figured out that yep yeet all the unicode to the sun 😅
`From:  some text <user@domain.com>`

Interesting part is that for some reason this same unicode doesn't error in the "Subject" header. I don't know why there is a diffrence like that.

Here is the error code I am getting when it is present in the "from" headers:

```
Response (if any):
{"name":"An Internal Error Has Occurred.","message":"An Internal Error Has Occurred.","url":"\/objects\/add\/4382"}
Traceback (most recent call last):
  File "C:\Users\samit\Documents\GitHub\PyMISP\examples\add_email_object.py", line 28, in <module>
    response = pymisp.add_object(args.event, eo, pythonify=True)
  File "C:\Users\samit\AppData\Local\Programs\Python\Python39\lib\site-packages\pymisp-2.4.148-py3.9.egg\pymisp\api.py", line 562, in add_object
    new_object = self._check_json_response(r)
  File "C:\Users\samit\AppData\Local\Programs\Python\Python39\lib\site-packages\pymisp-2.4.148-py3.9.egg\pymisp\api.py", line 3451, in _check_json_response
    r = self._check_response(response, expect_json=True)
  File "C:\Users\samit\AppData\Local\Programs\Python\Python39\lib\site-packages\pymisp-2.4.148-py3.9.egg\pymisp\api.py", line 3468, in _check_response
    raise MISPServerError(f'Error code 500:\n{response.text}')
pymisp.exceptions.MISPServerError: Error code 500:
{"name":"An Internal Error Has Occurred.","message":"An Internal Error Has Occurred.","url":"\/objects\/add\/4382"} 
```

I have also come across some others in different circumstances but unfortunately I haven't documented them back then. I have had trouble with some emojis when they have been in text and usually saved to the attribute comment section but I have no idea are those related to this one under the hood. Only common factor in my personal experience has been that if there are emojis in text they cause errors upon saving. Which could have been the reason to jump the gun here *cough*.

Also just curious to understand the differences here but do you guys see them as emojis or those question marks when you view them from the MISP event UI?

If you want we can scrap that unicode to ascii part from the code and continue on with a separate issue on that emoji problem? If you have any thoughts on where to open the issue please do let me know I don't really have a clue why it is doing that and where the root cause is exactly?

Rafiot commented:
First of all I see you're on windows and I'm glad to see PyMISP works, that's not something I was expecting, if you have any issue, please let me know, and PRs are welcome.

Okay, so back to the problem at hand. Encoding (and emails) are a can of worms and quite a few things can be happening, and I think you have the following cases:

1. the mail is proper utf-8, the emoji is UTF-8, python does its thing, creates the MISP Object, posts it, MISP takes it, displays everything as it should, all good
2. the mail is in something else than UTF-8 but the python module figures it out and we fall back to `1`
3. the mail is in something else than UTF-8 and the python module doesn't know what to, tries to decode, fails, and turns it into `?` 
4. Then, the actual exception (500) back from MISP shouldn't happen, and I'd love to figure it out more in details. Is it possible for you to look at the errors logs on the machine? They are located in `<MISP>/app/tmp/logs/error.log` and you should be able to get the stacktrace which will give more information about the issue.

And let's merge the PR without decoding until we understand more precisely what's going on.

samitainio commented:
@Rafiot 

I'll document it here for now.

I created a test event normally throught the UI and added this attribute:
- External analysis, text: 💋
- Results in ? showing up in the UI but it got saved.
- Then I added a comment to the same attribute: 💋 which then resulted in the error below:

```
==> /var/www/MISP/app/tmp/logs/error.log <==
2021-09-28 17:22:05 Error: [PDOException] SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect string value: '\xF0\x9F\x92\x8B' for column `misp`.`attributes`.`comment` at row 1
Request URL: /attributes/editField/10869860
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/Database/Mysql.php(425): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1926): Mysql->update()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1760): Model->_doSave()
#5 /var/www/MISP/app/Controller/AttributesController.php(963): Model->save()
#6 [internal function]: AttributesController->editField()
#7 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(499): ReflectionMethod->invokeArgs()
#8 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction()
#9 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#10 /var/www/MISP/app/webroot/index.php(101): Dispatcher->dispatch()
#11 {main}
```

Then I also did try to add the problematic email again:

```
2021-09-28 17:31:04 Error: [PDOException] SQLSTATE[22007]: Invalid datetime format: 1366 Incorrect string value: '\xF0\x9F\x92\x8B s...' for column `misp`.`attributes`.`comment` at row 1
Request URL: /objects/add/4382
Stack Trace:
#0 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(502): PDOStatement->execute()
#1 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(468): DboSource->_execute()
#2 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Datasource/DboSource.php(1132): DboSource->execute()
#3 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1942): DboSource->create()
#4 /var/www/MISP/app/Lib/cakephp/lib/Cake/Model/Model.php(1760): Model->_doSave()
#5 /var/www/MISP/app/Model/Attribute.php(3262): Model->save()
#6 /var/www/MISP/app/Model/MispObject.php(469): Attribute->saveAttributes()
#7 /var/www/MISP/app/Controller/ObjectsController.php(272): MispObject->saveObject()
#8 [internal function]: ObjectsController->add()
#9 /var/www/MISP/app/Lib/cakephp/lib/Cake/Controller/Controller.php(499): ReflectionMethod->invokeArgs()
#10 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(193): Controller->invokeAction()
#11 /var/www/MISP/app/Lib/cakephp/lib/Cake/Routing/Dispatcher.php(167): Dispatcher->_invoke()
#12 /var/www/MISP/app/webroot/index.php(101): Dispatcher->dispatch()
#13 {main}
```

Rafiot commented:
the `Invalid datetime format` is extremely weird and should definitely not happen. I passed it to my colleagues.

Rafiot commented:
Okay, so I got a potential reason: what is the format of your database? If the tables aren't `utf8mb4`, MySQL doesn't accept 4 character utf representation

samitainio commented:
@Rafiot

Related variables:
```
MariaDB [(none)]> show variables where variable_name like "character_set%" or variable_name like "collation_%";
+--------------------------+----------------------------+
| Variable_name            | Value                      |
+--------------------------+----------------------------+
| character_set_client     | utf8mb4                    |
| character_set_connection | utf8mb4                    |
| character_set_database   | utf8mb4                    |
| character_set_filesystem | binary                     |
| character_set_results    | utf8mb4                    |
| character_set_server     | utf8mb4                    |
| character_set_system     | utf8mb3                    |
| character_sets_dir       | /usr/share/mysql/charsets/ |
| collation_connection     | utf8mb4_general_ci         |
| collation_database       | utf8mb4_unicode_ci         |
| collation_server         | utf8mb4_unicode_ci         |
+--------------------------+----------------------------+
```

Database wide settings:
```
MariaDB [misp]> select schema_name, default_character_set_name, default_collation_name from information_schema.schemata where schema_name = "misp";
+-------------+----------------------------+------------------------+
| schema_name | default_character_set_name | default_collation_name |
+-------------+----------------------------+------------------------+
| misp        | utf8mb4                    | utf8mb4_unicode_ci     |
+-------------+----------------------------+------------------------+
```
Table collation:
```
MariaDB [(none)]> select table_schema, table_name, table_collation from information_schema.tables where table_schema="misp" and table_name="attributes" group by table_name;
+--------------+------------+--------------------+
| table_schema | table_name | table_collation    |
+--------------+------------+--------------------+
| misp         | attributes | utf8mb4_unicode_ci |
+--------------+------------+--------------------+

```

Table misp.attributes:
```
MariaDB [(none)]> show full columns from misp.attributes;
+---------------------+--------------+--------------------+------+-----+---------+----------------+---------------------------------+---------+
| Field               | Type         | Collation          | Null | Key | Default | Extra          | Privileges                      | Comment |
+---------------------+--------------+--------------------+------+-----+---------+----------------+---------------------------------+---------+
| id                  | int(11)      | NULL               | NO   | PRI | NULL    | auto_increment | select,insert,update,references |         |
| event_id            | int(11)      | NULL               | NO   | MUL | NULL    |                | select,insert,update,references |         |
| object_id           | int(11)      | NULL               | NO   | MUL | 0       |                | select,insert,update,references |         |
| object_relation     | varchar(255) | utf8mb4_bin        | YES  | MUL | NULL    |                | select,insert,update,references |         |
| category            | varchar(255) | utf8mb4_bin        | YES  | MUL | NULL    |                | select,insert,update,references |         |
| type                | varchar(100) | utf8mb4_unicode_ci | YES  | MUL | NULL    |                | select,insert,update,references |         |
| value1              | mediumtext   | utf8mb4_unicode_ci | YES  | MUL | NULL    |                | select,insert,update,references |         |
| value2              | mediumtext   | utf8mb4_unicode_ci | YES  | MUL | NULL    |                | select,insert,update,references |         |
| to_ids              | tinyint(1)   | NULL               | NO   |     | 1       |                | select,insert,update,references |         |
| uuid                | varchar(40)  | utf8mb4_bin        | YES  | UNI | NULL    |                | select,insert,update,references |         |
| timestamp           | int(11)      | NULL               | NO   |     | 0       |                | select,insert,update,references |         |
| distribution        | tinyint(4)   | NULL               | NO   |     | 0       |                | select,insert,update,references |         |
| sharing_group_id    | int(11)      | NULL               | NO   | MUL | NULL    |                | select,insert,update,references |         |
| comment             | text         | utf8mb4_unicode_ci | YES  |     | NULL    |                | select,insert,update,references |         |
| deleted             | tinyint(1)   | NULL               | NO   |     | 0       |                | select,insert,update,references |         |
| disable_correlation | tinyint(1)   | NULL               | NO   |     | 0       |                | select,insert,update,references |         |
| first_seen          | bigint(20)   | NULL               | YES  | MUL | NULL    |                | select,insert,update,references |         |
| last_seen           | bigint(20)   | NULL               | YES  | MUL | NULL    |                | select,insert,update,references |         |
+---------------------+--------------+--------------------+------+-----+---------+----------------+---------------------------------+---------+
```

samitainio commented:
@Rafiot 

I made some progress! The comments are now accepting emojis on my machine and aren't producing error messages anymore. The table `misp.logs` column `change` needed to be edited with the following command:

```ALTER TABLE misp.logs MODIFY `change` TEXT CHARACTER SET utf8mb4 COLLATE utf8mb4_bin;```

However the value fields are still accepting emojis as input but after saving they show up as question marks. I am not getting any error messages to logs while adding them. I'll keep digging into that and post results here. I'll take any pointers and ideas if you have thoughts on where to look at.
